File: ContinenteScrapper.py
from datetime import date
import re
import string
from bs4 import BeautifulSoup, Tag
from Utils import make_request, pipe

class ContinenteScrapper():
    def __init__(self):
        self.url = "https://www.continente.pt/on/demandware.store/Sites-continente-Site/default/Search-UpdateGrid?cgid=col-produtos&pmin=0%2e01&start=0&sz=24"
        self.name = 'Continente'
        self.product_selector = '[data-pid].product'
        self.name_selector = '[data-pid].product .pwc-tile--description'
        self.price_selector = '[data-pid].product .js-product-price .value'
        self.price_selector_attr = 'content'
        self.price_unit_selector = '[data-pid].product .pwc-m-unit'
    
    def scrape(self):
        return pipe(self.url).apply(make_request).apply(self.extract_products).value

    def extract_products(self, response):
        soup = BeautifulSoup(response.text, 'html.parser')
        products = soup.select(self.product_selector)

        def get_price(htmlelement: Tag):
            if self.price_selector_attr:
                return htmlelement.attrs.get(self.price_selector_attr).strip().replace(',', '.')
            else:
                return htmlelement.text.strip().replace(',', '.')
        
        def get_price_element(product: Tag):
            return product.select_one(self.price_selector)
        
        def get_price_unit(product: Tag) -> string:
            text = product.select_one(self.price_unit_selector).text

            return re.findall(r"([a-zA-Z]+)",text)[0]

        return [{
            'Name': product.select_one(self.name_selector).text.strip(),
            'Price': float(pipe(product).apply(get_price_element).apply(get_price).value),
            'PriceUnit': get_price_unit(product).lower(),
            'Source': self.name,
            'Date': str(date.today())
        } for product in products]

File: ContinenteScrapper.fs
module GNogueira.SupermarketScrapeTool.Service.ContinenteScrapper
open System
open System.Collections.Generic
open System.Net
open System.Net.Http
open System.Text.RegularExpressions
open HtmlAgilityPack
open Fizzler.Systems.HtmlAgilityPack
open FSharpPlus
open Models

let pageStart = 0
let pageSize = 999999
let supermarketUrl = $"https://www.continente.pt/on/demandware.store/Sites-continente-Site/default/Search-UpdateGrid?cgid=col-produtos&pmin=0%%2e01&start={pageStart}&sz={pageSize}"
let supermarketName = Continente
let productSelector = "[data-pid].product"
let nameSelector = ".pwc-tile--description"
let priceSelector = ".js-product-price .value"
let priceSelectorAttr = "content"
let priceUnitSelector = ".pwc-m-unit"

let makeRequest (url: string) =
    async {
        use httpClient = new HttpClient()
        
        let! response = httpClient.GetAsync(url) |> Async.AwaitTask
        response.EnsureSuccessStatusCode() |> ignore
        
        let! content = response.Content.ReadAsStringAsync() |> Async.AwaitTask
        
        return content
    } |> Async.RunSynchronously


let getPriceUnit (product: HtmlNode) =
    let matchText pattern text = Regex.Match(text, pattern)
    let getGroup (id: int) (matchObj: Match) = matchObj.Groups[id].Value
    let toLower (text: string) = text.ToLower()
    
    product.InnerText
    |> matchText "([a-zA-Z]+)"
    |> getGroup 1
    |> toLower

let scrape() =
    let getNode selector (node : HtmlNode) = node.QuerySelector(selector)
    let getName (node: HtmlNode) = node.InnerText |> String.trimWhiteSpaces
    let getPrice (node: HtmlNode) =
        match priceSelectorAttr with
        | "" -> node.InnerText.Trim().Replace(",", ".")
        | attr -> node.GetAttributeValue(attr, "").Trim().Replace(",", ".")
        |> float
    
    supermarketUrl
    |> makeRequest
    |> fun response ->
        let loadHtml resp (htmlDoc : HtmlDocument) =
            htmlDoc.LoadHtml(resp)
            htmlDoc
        
        HtmlDocument() |> loadHtml response
    
    |> fun doc ->
        doc.DocumentNode.QuerySelectorAll(productSelector)
        |> Seq.map (fun (product : HtmlNode) ->
            let name = product |> getNode nameSelector |> getName
            let price = product |> getNode priceSelector |> getPrice
            let priceUnit = product |> getNode priceUnitSelector |> getPriceUnit |> PriceUnit.ofString
            let currentDate = DateTime.Now.ToString("yyyy-MM-dd")
            
            {id = Guid.NewGuid();  Name = name; Price = price; PriceUnit = priceUnit; Source = supermarketName; Date = currentDate }
        )